### PR TITLE
Implement `Folder` trait for `Trace` to improve compatibility with folding types

### PR DESCRIPTION
**Pull request by querolita**
This could be used inside the `fold()` in `msm/proof.rs` as well as in tests.

### CODE REVIEW

**Comment by dannywillems** [comment]
I prefer this idea than the one I was trying in https://github.com/o1-labs/proof-systems/pull/2164. I will continue with these changes @querolita

File: optimism/src/trace.rs [hunk 1]
```diff
@@ -75,6 +83,66 @@ impl<
...
+            .witness
+            .cols
+            .iter()
+            .map(|w| srs.commit_evaluations_non_hiding(domain, w))
+            .map(|c| c.elems[0])
+            .collect_vec()
+            .try_into()
+            .unwrap();
+
+        // FIXME: this would need the sponge instead to obtain the challenges from
```
Comment (dannywillems) [inline optimism/src/trace.rs:130]: I'm taking care of this now. Will be in another PR.

File: folding/src/checker.rs [hunk 2]
```diff
@@ -0,0 +1,342 @@
+//! A kind of pseudo-prover, will compute the expressions over the witness a check row by row
```
Comment (dannywillems) [inline folding/src/checker.rs:1]: Is it expected in this PR?

File: optimism/src/trace.rs (hunk 1 above)
Comment (dannywillems) [reply optimism/src/trace.rs:130]: It seems that the randomness is not carefully handled in the folding library. It requires to change the library (see conversation on Slack). Taking care of it.

File: folding/src/checker.rs (hunk 2 above)
Comment (querolita) [reply folding/src/checker.rs:1]: Oh shoot, why is this commit in this PR... No, this is meant to be in https://github.com/o1-labs/proof-systems/pull/2181

File: optimism/src/trace.rs (hunk 1 above)
Comment (querolita) [reply optimism/src/trace.rs:130]: Will fix it, give me a minute

**Review by dannywillems — APPROVED** [review]
